File: articles/prototyping-design-thinking/go/prototype_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type Prototype struct {
	Name                    string
	LearningGain            float64
	FeasibilitySignal       float64
	UserResponse            float64
	EquityValue             float64
	ImplementationRelevance float64
	EthicalRisk             float64
	OperationalRisk         float64
	TechnicalRisk           float64
	ScalingRisk             float64
	CompositeRisk           float64
	PrototypeValue          float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func compositeRisk(p Prototype) float64 {
	return 0.30*p.EthicalRisk +
		0.30*p.OperationalRisk +
		0.20*p.TechnicalRisk +
		0.20*p.ScalingRisk
}

func prototypeValue(p Prototype) float64 {
	return 0.25*p.LearningGain +
		0.18*p.FeasibilitySignal +
		0.20*p.UserResponse +
		0.15*p.EquityValue +
		0.12*p.ImplementationRelevance -
		0.10*p.CompositeRisk
}

func loadPortfolio(path string) []Prototype {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var prototypes []Prototype

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 12 {
			log.Fatalf("malformed row: %v", row)
		}

		p := Prototype{
			Name:                    row[0],
			LearningGain:            parseFloat(row[3]),
			FeasibilitySignal:       parseFloat(row[4]),
			UserResponse:            parseFloat(row[5]),
			EquityValue:             parseFloat(row[6]),
			ImplementationRelevance: parseFloat(row[7]),
			EthicalRisk:             parseFloat(row[8]),
			OperationalRisk:         parseFloat(row[9]),
			TechnicalRisk:           parseFloat(row[10]),
			ScalingRisk:             parseFloat(row[11]),
		}
		p.CompositeRisk = compositeRisk(p)
		p.PrototypeValue = prototypeValue(p)
		prototypes = append(prototypes, p)
	}

	return prototypes
}

func main() {
	path := "../data/raw/prototype_portfolio_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	prototypes := loadPortfolio(path)

	sort.Slice(prototypes, func(i, j int) bool {
		return prototypes[i].PrototypeValue > prototypes[j].PrototypeValue
	})

	fmt.Println("rank,prototype,prototype_value,composite_risk")
	for i, p := range prototypes {
		fmt.Printf("%d,%s,%.6f,%.6f\n", i+1, p.Name, p.PrototypeValue, p.CompositeRisk)
	}
}

File: articles/prototyping-design-thinking/c/prototype_value_engine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_PROTOTYPES 128
#define MAX_LINE 512
#define MAX_NAME 180

typedef struct {
    char name[MAX_NAME];
    double learning_gain;
    double feasibility_signal;
    double user_response;
    double equity_value;
    double implementation_relevance;
    double ethical_risk;
    double operational_risk;
    double technical_risk;
    double scaling_risk;
    double composite_risk;
    double prototype_value;
} Prototype;

double compute_composite_risk(const Prototype *p) {
    return 0.30 * p->ethical_risk +
           0.30 * p->operational_risk +
           0.20 * p->technical_risk +
           0.20 * p->scaling_risk;
}

double compute_prototype_value(const Prototype *p) {
    return 0.25 * p->learning_gain +
           0.18 * p->feasibility_signal +
           0.20 * p->user_response +
           0.15 * p->equity_value +
           0.12 * p->implementation_relevance -
           0.10 * p->composite_risk;
}

int compare_desc(const void *a, const void *b) {
    const Prototype *pa = (const Prototype *)a;
    const Prototype *pb = (const Prototype *)b;

    if (pa->prototype_value < pb->prototype_value) return 1;
    if (pa->prototype_value > pb->prototype_value) return -1;
    return 0;
}

int main(int argc, char **argv) {
    const char *path = argc > 1 ? argv[1] : "../data/raw/prototype_portfolio_raw.csv";
    FILE *file = fopen(path, "r");

    if (!file) {
        fprintf(stderr, "ERROR: Unable to open input file: %s\n", path);
        return 1;
    }

    Prototype prototypes[MAX_PROTOTYPES];
    char line[MAX_LINE];
    int count = 0;

    fgets(line, sizeof(line), file); /* header */

    while (fgets(line, sizeof(line), file) && count < MAX_PROTOTYPES) {
        char *token = strtok(line, ",");
        if (!token) continue;

        strncpy(prototypes[count].name, token, MAX_NAME - 1);
        prototypes[count].name[MAX_NAME - 1] = '\0';

        strtok(NULL, ","); /* prototype_type */
        strtok(NULL, ","); /* fidelity_level */

        token = strtok(NULL, ",");
        prototypes[count].learning_gain = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        prototypes[count].feasibility_signal = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        prototypes[count].user_response = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        prototypes[count].equity_value = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        prototypes[count].implementation_relevance = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        prototypes[count].ethical_risk = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        prototypes[count].operational_risk = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        prototypes[count].technical_risk = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        prototypes[count].scaling_risk = token ? atof(token) : 0.0;

        prototypes[count].composite_risk = compute_composite_risk(&prototypes[count]);
        prototypes[count].prototype_value = compute_prototype_value(&prototypes[count]);

        count++;
    }

    fclose(file);

    qsort(prototypes, count, sizeof(Prototype), compare_desc);

    printf("rank,prototype,prototype_value,composite_risk\n");
    for (int i = 0; i < count; i++) {
        printf("%d,%s,%.6f,%.6f\n", i + 1, prototypes[i].name, prototypes[i].prototype_value, prototypes[i].composite_risk);
    }

    return 0;
}
